cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/ReportGenerationSpark.java

```diff
@@ -137,17 +158,14 @@ public void getReports(HttpServiceRequest request, HttpServiceResponder responde
       while (idx < reportIdDirs.size() && reportStatuses.size() < limit) {
         Location reportIdDir = reportIdDirs.get(idx++);
         String reportId = reportIdDir.getName();
-        try {
-          reportStatuses.add(new ReportStatusInfo(reportId, getReportMetaInfo(reportId, reportIdDir,
-                                                                              getReportRequest(reportIdDir))));
-        } catch (Exception e) {
-          LOG.error("Failed to get the status for report with id {}", reportId, e);
-          responder.sendError(500, String.format("Failed to get the status for report with id %s" +
-                                                   " because of error: %s", reportId, e.getMessage()));
-          return;
+        ReportMetaInfo metaInfo = getReportMetaInfo(reportId, reportIdDir, getReportRequest(reportIdDir));
+        // skip adding reports that are to be deleted
+        if (ReportStatus.DELETED.equals(metaInfo.getStatus())) {
+          continue;
         }
+        reportStatuses.add(new ReportStatusInfo(reportId, metaInfo));
       }
-      responder.sendJson(200, new ReportList(offset, limit, reportIdDirs.size(), reportStatuses));
+      return new ReportList(offset, limit, reportIdDirs.size(), reportStatuses);
```

Comment:
using offset blindly might cause duplicate results to be sent. lets say initially offset is 0 and limit is 10, but say report 1-5 is expired and you return the reports from 6-15 in the response. if the user will send the next query with offset 10 and limit 10, the reports from 11-15 will also be sent. which is incorrect, the correct request should have offset as 16, which can be known from previous query if the response had `idx` information.  another option is to improve the logic. where offset means skip non-expired records. offset of 10 would mean skip first 10 non-expired records, which i think is better. but this issue has to be addressed.